clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -560,12 +561,14 @@ private List<ProducerBatch> drainBatchesForOneNode(Cluster cluster, Node node, i
         List<PartitionInfo> parts = cluster.partitionsForNode(node.id());
         List<ProducerBatch> ready = new ArrayList<>();
         /* to make starvation less likely this loop doesn't start at 0 */
+        int drainIndex = getDrainIndex(node.idString());
         int start = drainIndex = drainIndex % parts.size();
+        updateDrainIndex(node.idString(),drainIndex);
         do {
             PartitionInfo part = parts.get(drainIndex);
             TopicPartition tp = new TopicPartition(part.topic(), part.partition());
-            this.drainIndex = (this.drainIndex + 1) % parts.size();
-
+            drainIndex = (drainIndex + 1) % parts.size();
+            updateDrainIndex(node.idString(),drainIndex);
```

Comment:
Same here, don't need to update it each time.

Reply:
thanks for the advice , i will modify and resubmit